network/discovery/discovery.go

```diff
@@ -0,0 +1,427 @@
+package discovery
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/0xPolygon/polygon-edge/network/common"
+	"github.com/0xPolygon/polygon-edge/network/event"
+	"github.com/hashicorp/go-hclog"
+	"github.com/libp2p/go-libp2p-core/network"
+	"time"
+
+	"github.com/0xPolygon/polygon-edge/network/grpc"
+	"github.com/0xPolygon/polygon-edge/network/proto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	kb "github.com/libp2p/go-libp2p-kbucket"
+)
+
+const (
+	// maxDiscoveryPeerReqCount is the max peer count that
+	// can be requested from other peers
+	maxDiscoveryPeerReqCount = 16
+
+	// peerDiscoveryInterval is the interval at which other
+	// peers are queried for their peer sets
+	peerDiscoveryInterval = 5 * time.Second
+
+	// bootnodeDiscoveryInterval is the interval at which
+	// random bootnodes are dialed for their peer sets
+	bootnodeDiscoveryInterval = 60 * time.Second
+)
+
+// networkingServer defines the base communication interface between
+// any networking server implementation and the DiscoveryService
+type networkingServer interface {
+	// BOOTNODE QUERIES //
+
+	// IsBootnode checks if the peer is a registered bootnode
+	IsBootnode(peerID peer.ID) bool
+
+	// GetRandomBootnode fetches a random bootnode, if any
+	GetRandomBootnode() *peer.AddrInfo
+
+	// GetBootnodeConnCount fetches the number of bootnode connections [Thread safe]
+	GetBootnodeConnCount() int64
+
+	// PROTOCOL MANIPULATION //
+
+	// NewDiscoveryClient returns a discovery gRPC client connection
+	NewDiscoveryClient(peerID peer.ID) (proto.DiscoveryClient, error)
```

Comment:
Because it allows us to easily mock the gRPC client interface for the discovery service, and because the base networking server is in charge of keeping track of stream / peer connections. The discovery service needs a reference to an _existing_ stream if it's present, and the only way to get it is from the base networking server.

It's just about shifting the responsibility of connection management - discovery, like identity, are just services that run on top of the base networking server, and they should be kept minimal